polynote-kernel/src/main/scala/polynote/kernel/interpreter/scal/SemanticDbScan.scala

```diff
@@ -70,20 +116,47 @@ class SemanticDbScan (compiler: ScalaCompiler) {
     }.map(_.flatten)
   }
 
-  private def indexSources(sources: List[SourceFile]) = for {
-    completed <- Ref.make(0)
-    _         <- ZIO.foreachPar_(sources) {
-      sourceFile => for {
-        _            <- index(sourceFile).retryN(3).unit.catchAll {
-          cause =>
-            Logging.error(s"Error indexing ${sourceFile.file.name}", cause)
-        }
-        numCompleted <- completed.updateAndGet(_ + 1)
-        _            <- CurrentTask.setProgress(numCompleted.toDouble / sources.size)
-      } yield ()
+  private def indexSources(sources: List[SourceFile]) = ZIO {
+    // TODO: can use the profiler: scala.tools.nsc.profile.Profiler to track detailed progress here.
+    val units = sources.filter {
+      sourceFile => sourceFile.path.split('.').last match {
+        case "java" | "scala" => true
+        case _                => false
+      }
+    }.map {
+      file =>
+        val unit = new semanticdbGlobal.RichCompilationUnit(file)
+        compileUnits.put(file.file, unit)
+        if (file.file.name.endsWith(".java"))
+          javaMapping.put(file.path.split("!").last.stripSuffix(".java").replace('/', '.'), unit)
+        unit
     }
-  } yield ()
+    // do an initial namer run of all the sources together
+    run.compileUnits(units, run.parserPhase)
+    units
+  }.flatMap {
+    units =>
+      // if there were errors doing that, it could result in some sources getting skipped because of other bad sources.
+      // so now loop through them all again, running parser and namer on each one individually.
+      Ref.make(0).flatMap {
+        completed =>
+          ZIO.foreach_(units) {
+            unit => ZIO {
+              run.parserPhase.asInstanceOf[semanticdbGlobal.GlobalPhase].apply(unit)
+              run.namerPhase.asInstanceOf[semanticdbGlobal.GlobalPhase].apply(unit)
+            } *> completed.updateAndGet(_ + 1).flatMap {
+              numCompleted => CurrentTask.setProgress(numCompleted.toDouble / sources.size)
+            }
+          }
+      }.zipLeft(ZIO {
+        semanticdbGlobal.settings.stopAfter.value = List("typer")
+        run = new semanticdbGlobal.Run()
+      })
+  }
 
 }
 
+object SemanticDbScan {
```

Comment:
unnecessary object?

Reply:
Yeah, I need to put an effectful constructor in here.